ds701_book/27-RNN.qmd

```diff
@@ -0,0 +1,640 @@
+
+- **Sequential Data**: Ideal for tasks where data points are dependent on previous ones.
+- **Memory**: Capable of retaining information from previous inputs, making them powerful for tasks involving context. This is achieved from hidden states and feedback loops.
+
+## RNN Applications
+
+- **Natural Language Processing (NLP)**: Language translation, sentiment analysis, and text generation.
+- **Speech Recognition**: Converting spoken language into text.
+- **Time Series Prediction**: Forecasting stock prices, weather, and other temporal data.
+- **Music Generation**: Creating a sequence of musical notes.
+
+## Outline
+
+1. Basic RNN Architecture
+1. LSTM (Long Short-Term Memory) and GRU (Gated Recurrent Units)
+1. Examples
+
+## RNN Architecture
+
+![](drawio/RNN-Full.png)
+
+## Forward Propagation
+
+The forward pass in the RNN architecture is given by the following operations
+
+1. $h_t = g_h(W_{hh} h_{t-1} + W_{hx} x_t + b_h)$
+1. $y_t = g_y(W_{yh} h_t + b_y)$
+
+The vector $x_t$ is the $t$-th element of the input sequence. The vector $h_t$ is the hidden state at the $t$-th point of the sequence. The vector $y_t$ is the $t$-th output of the sequence. The functions $g_h$ and $g_y$ are nonlinear activation functions. The weight matrices $W_{hh}$, $W_{hx}$, $W_{yh}$, and biases $b_h$, $b_y$ are trainable parameters. Note that we must define the vector $h_0$. A common choice is $h_0 = 0$.
+
+## Model Types
+
+Depending on the application there many be varying number of outputs $y_t$.
+
+:::: {.columns}
+::: {.column width="60%"}
+![](figs/RNNConfigurations.png)
+:::
+::: {.column width="40%"}
+1. One-to-one, is a regular feed forward neural network.
+1. One-to-many, e.g., be used for image captioning (1 image and output a sequence of words).
+1. Many-to-one, e.g., sentiment analysis from a sequence of words or stock price prediction.
+1. Many-to-many, e.g., machine translation or video frame-by-frame action classification.
+:::
+::::
+
+## Loss Calculation for Sequential Data
+
+For each prediction in the sequence $y_t$ and corresponding true value $\hat{y}_t$  we calculate the the cross-entropy loss $\mathcal{L}_t(\hat{y}_t, y_t) = -\hat{y}_t\log(y_t)$. The total loss is the sum of all the individual losses, $\mathcal{L} = \sum_t \mathcal{L}_t$. This is illustrated in the figure below.
+
+If we were predicting continious value we would use a mean-square error loss.
+
+![](drawio/RNN-Loss.png)
+
+
+## Vanishing Gradients
+
+The key idea of RNNs is that earlier items in the sequence  influence the more recent outputs. This is highlighted in blue in the figure below.
+
+However, for longer sequences the influence can be significantly reduced. This is illustrated in red in the figure below.
+
+Due to the potential for long sequences of data, during training you are likely to encounter small gradients. As a result, RNNs are prone to vanishing gradients.
+
+![](drawio/VanishingGradient.png)
+
+## RNN Limitations
+
+- **Vanishing Gradients**:
```

Comment:
i didn't double check, but is it possible to have exploding gradients too?

Reply:
Yes, I believe both are possible though I think that vanishing gradients are more common.